crates/libp2p-networking/src/network/behaviours/dht/cache.rs

```diff
@@ -0,0 +1,322 @@
+                    self.cache.insert(key.clone(), value);
+                    self.expiries.write().await.insert(expiry, key);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn save(&self) -> Result<(), CacheError> {
+        if let Some(filename) = &self.config.filename {
+            // prune first
+            self.prune().await;
+
+            // serialize
+            let mut cache_to_write = HashMap::new();
+            let expiries = self.expiries.read().await;
+            for (expiry, key) in &*expiries {
+                if let Some(entry) = self.cache.get(key) {
+                    cache_to_write.insert(expiry, (key, entry.value().clone()));
+                } else {
+                    tracing::warn!("key not found in cache: {:?}", key);
+                    Err(CacheError::GeneralCache {
+                        source: Box::new(bincode::ErrorKind::Custom(
+                            "key not found in cache".to_string(),
+                        )),
+                    })?;
+                };
+            }
+
+            let encoded = bincode_opts()
+                .serialize(&cache_to_write)
+                .context(SerializationSnafu)?;
+
+            std::fs::write(filename, encoded).context(DiskSnafu)?;
+        }
+
+        Ok(())
+    }
+
+    async fn prune(&self) {
+        let now = SystemTime::now();
+        let mut expiries = self.expiries.write().await;
+        let mut removed: u32 = 0;
+
+        while let Some((expires, key)) = expiries.pop_first() {
+            if now > expires {
+                self.cache.remove(&key);
+                removed += 1;
+            } else {
+                expiries.insert(expires, key);
+                break;
+            }
+        }
+
+        // todo: test speed on this as opposed to inline
+        if removed > 0 {
+            self.disk_parity_delta.fetch_add(removed, Ordering::Relaxed);
+        }
+    }
+
+    pub async fn get(&self, key: &Vec<u8>) -> Option<Ref<'_, Vec<u8>, Vec<u8>>> {
+        // prune, save if necessary
+        self.prune().await;
+        self.save_if_necessary().await;
+
+        // get
+        self.cache.get(key)
+    }
+
+    pub async fn insert(&self, key: Vec<u8>, value: Vec<u8>) {
+        // insert into cache and expiries
+        self.cache.insert(key.clone(), value);
+        self.expiries
+            .write()
+            .await
+            .insert(SystemTime::now() + self.config.expiry, key);
+
+        // save if reached max disk parity delta
+        self.disk_parity_delta.fetch_add(1, Ordering::Relaxed);
+        self.save_if_necessary().await;
+    }
+
+    async fn save_if_necessary(&self) {
+        let cur_disk_parity_delta = self.disk_parity_delta.load(Ordering::Relaxed);
+        if cur_disk_parity_delta >= self.config.max_disk_parity_delta {
+            if let Err(err) = self.save().await {
+                tracing::warn!("failed to save cache to file: {}", err);
```

Comment:
Might be worth making this an error. If it's not able to save the cache to a file, there's likely some sort of configuration failure going on (or the disk is out of space). Either way, loud failure seems good.